File: 0035-search-insert-position/0035-search-insert-position.py
# time complexity: O(logn)
# space complexity: O(1)
from typing import List


class Solution:
    def searchInsert(self, nums: List[int], target: int) -> int:
        left = 0
        right = len(nums)
        while left < right:
            mid = left + (right - left) // 2
            if nums[mid] > target:
                right = mid
            elif nums[mid] < target:
                left = mid + 1
            else:
                return mid
        return left


nums = [1, 3, 5, 6]
target = 2
print(Solution().searchInsert(nums, target))

File: C++/0035-search-insert-position.cpp
class Solution
{
public:
    int searchInsert(vector<int> &nums, int target)
    {
        int index = 0;
        if (nums.size() == 1)
            return (nums[0] < target ? 1 : 0);
        for (int i = 0; i < nums.size(); i++)
        {
            if (nums[i] == target)
                return i;
            if (nums[i] < target && nums[i + 1] > target)
                return i + 1;
            if (nums[nums.size() - 1] < target)
                return nums.size();
        }
        return index;
    }
};